### PR TITLE
Improve top & bottom bars for navigation

### PR DESCRIPTION
**Pull request by dishanest**
## Overview

Addressed https://github.com/cuappdev/volume-ios/issues/82, made BrowserView more closely adhere to [Figma designs](https://www.figma.com/file/nWE6pUB1O42nzaXPGj1uMt/?node-id=121%3A0), and fixed various other issues.
## Changes Made

- Implemented a floating tab bar over the default bar provided by `TabView` that can't be configured at all. This floating bar can be configured to look like any other view.
- `BrowserView`:
  - Implemented custom nav bar
    - Follows original designs closely w/ back and "open in Safari" buttons
    - Shrinks when WebView scrolls down
    - Drop shadow
  - Hide the bottom toolbar when scrolling down
- Moved `NavigationLink`s to `BrowserView` from being embedded within `Row` structs to being embedded in their navigating source view. This better follows the paradigm of navigation being directed by ancestor views rather than lower-level views.
- Fixed bug where simulator couldn't get past onboarding due to `deviceToken` not being set on simulators. It is now possible to test on non-physical devices. (Notifications still don't work on simulators of course)

## Test Coverage

Play tested on iPhone 8, iPhone 8 Plus, iPhone 13 Pro. Verified onboarding, reading, & other tabs look fine. Font does not appear to be too small on a physical device. 

## Next Steps

Merge this PR first, then build magazines off of this.

## Screenshots 

<details>

  <summary> iPhone 8 </summary>

Simulator Screen Shot - iPhone 8 - 2022-03-09 at 04 30 19
Simulator Screen Shot - iPhone 8 - 2022-03-09 at 04 30 55
Simulator Screen Shot - iPhone 8 - 2022-03-09 at 04 31 04

</details>

<details>

  <summary> iPhone 8 Plus </summary>

Simulator Screen Shot - iPhone 8 Plus - 2022-03-09 at 04 32 17
Simulator Screen Shot - iPhone 8 Plus - 2022-03-09 at 04 32 31
Simulator Screen Shot - iPhone 8 Plus - 2022-03-09 at 04 32 36

</details>

<details>

  <summary> iPhone 13 Pro </summary>

Simulator Screen Shot - iPhone 13 Pro - 2022-03-09 at 04 31 31
Simulator Screen Shot - iPhone 13 Pro - 2022-03-09 at 04 31 47
Simulator Screen Shot - iPhone 13 Pro - 2022-03-09 at 04 31 53

</details>

### CODE REVIEW

File: Volume/Supporting Views/TabBarReader.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+struct TabBarReader: UIViewControllerRepresentable {
+    var completion: (UITabBar) -> Void
+    private let proxy = ProxyController()
+
+    func makeUIViewController(context: UIViewControllerRepresentableContext<TabBarReader>) -> UIViewController {
+        proxy.completion = completion
+        return proxy
+    }
+
+    func updateUIViewController(_ uiViewController: UIViewController, context: UIViewControllerRepresentableContext<TabBarReader>) { }
```
Comment (amyqxhuang) [inline Volume/Supporting Views/TabBarReader.swift:20]: is this supposed to be empty? can we remove this if it is?

File: Volume/Views/BrowserView.swift [hunk 2]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+                            Image("compass")
+                                .resizable()
+                                .aspectRatio(contentMode: .fit)
+                                .frame(height: 24)
+                                .foregroundColor(Color.volume.lightGray)
+                        }
+                    }
+                    .padding(.horizontal, 16)
+                    
+                    VStack(spacing: 0) {
```
Comment (amyqxhuang) [inline Volume/Views/BrowserView.swift:95]: I'm not sure but if spacing is 0, can we remove `(spacing: 0)`?

File: Volume/Views/MainView/MainView.swift [hunk 3]
```diff
@@ -12,35 +12,74 @@ import SwiftUI
...
         TabView(selection: $selectedTab) {
             TabContainer(screen: .homeList) {
                 HomeList()
             }
-            .tabItem {
-                Image("volume")
-            }
             .tag(Tab.home)
+            .background(TabBarReader{ tabBar in
+                print(tabBar.bounds.height)
```
Comment (amyqxhuang) [inline Volume/Views/MainView/MainView.swift:26]: remove print?

**Review by amyqxhuang — APPROVED** [review]
looks good to me 🤩🚀

File: Volume/Supporting Views/TabBarReader.swift (hunk 1 above)
Comment (Npikielny) [reply Volume/Supporting Views/TabBarReader.swift:20]: Can't remove bc it's from the protocol

File: Volume/Views/BrowserView.swift [hunk 4]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
         }
     }
     
+    private var navigationTitle: String {
+        switch state {
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:38]: Could you just make state an optional?

File: Volume/Views/BrowserView.swift (hunk 2 above)
Comment (dishanest) [reply Volume/Views/BrowserView.swift:95]: This isn't always the case but it works here.

File: Volume/Views/BrowserView.swift [hunk 5]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+                if showToolbars {
+                    HStack {
+                        Button {
+                            presentationMode.wrappedValue.dismiss()
+                        } label: {
+                            Image("arrow-left")
+                                .resizable()
+                                .aspectRatio(contentMode: .fit)
+                                .frame(height: 24)
+                                .foregroundColor(Color.black)
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:72]:
```suggestion
                                .foregroundColor(.black)
```

File: Volume/Views/BrowserView.swift [hunk 6]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+                        
+                        Spacer()
+                        
+                        if case let .results(article) = state, let url = article.articleUrl {
+                            Link(destination: url) {
+                                Image("compass")
+                                    .resizable()
+                                    .aspectRatio(contentMode: .fit)
+                                    .frame(height: 24)
+                                    .foregroundColor(Color.black)
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:83]:
```suggestion
                                    .foregroundColor(.black)
```

File: Volume/Views/BrowserView.swift [hunk 7]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+                            }
+                        } else {
+                            Image("compass")
+                                .resizable()
+                                .aspectRatio(contentMode: .fit)
+                                .frame(height: 24)
+                                .foregroundColor(Color.volume.lightGray)
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:90]:
```suggestion
                                .foregroundColor(.volume.lightGray)
```

File: Volume/Views/BrowserView.swift [hunk 8]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+                    }
+                    .padding(.horizontal, 16)
+                    
+                    VStack(spacing: 0) {
+                        Text(navigationTitle)
+                            .font(.begumBold(size: 12))
+                            .fixedSize()
+                        Text("Reading in Volume")
+                            .font(.latoRegular(size: 10))
+                            .foregroundColor(Color.volume.lightGray)
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:101]:
```suggestion
                            .foregroundColor(.volume.lightGray)
```

File: Volume/Views/BrowserView.swift [hunk 9]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+                                .aspectRatio(contentMode: .fit)
+                                .frame(height: 24)
+                                .foregroundColor(Color.volume.lightGray)
+                        }
+                    }
+                    .padding(.horizontal, 16)
+                    
+                    VStack(spacing: 0) {
+                        Text(navigationTitle)
+                            .font(.begumBold(size: 12))
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:97]: begum made my day, thank you.

File: Volume/Views/MainView/HomeList.swift [hunk 10]
```diff
@@ -162,8 +164,10 @@ struct HomeList: View {
                 }
             case .reloading(let results), .results(let results):
                 ForEach(results.followedArticles) { article in
-                    ArticleRow(article: article, navigationSource: .followingArticles)
-                        .padding([.leading, .trailing])
+                    NavigationLink(destination: BrowserView(initType: .readyForDisplay(article), navigationSource: .followingArticles)) {
+                        ArticleRow(article: article, navigationSource: .followingArticles)
+                            .padding([.leading, .trailing])
```
Comment (Npikielny) [inline Volume/Views/MainView/HomeList.swift:169]:
```suggestion
                            .padding(.horizontal)
```

File: Volume/Views/MainView/HomeList.swift [hunk 11]
```diff
@@ -182,8 +186,10 @@ struct HomeList: View {
                 Spacer().frame(height: 0)
             case .reloading(let results), .results(let results):
                 ForEach(results.otherArticles) { article in
-                    ArticleRow(article: article, navigationSource: .otherArticles)
-                        .padding([.bottom, .leading, .trailing])
+                    NavigationLink(destination: BrowserView(initType: .readyForDisplay(article), navigationSource: .otherArticles)) {
+                        ArticleRow(article: article, navigationSource: .otherArticles)
+                            .padding([.leading, .trailing])
```
Comment (Npikielny) [inline Volume/Views/MainView/HomeList.swift:191]:
```suggestion
                            .padding(.horizontal)
```

File: Volume/Views/MainView/MainView.swift [hunk 12]
```diff
@@ -12,35 +12,74 @@ import SwiftUI
...
+    private var tabViewContainer: some View {
         TabView(selection: $selectedTab) {
             TabContainer(screen: .homeList) {
                 HomeList()
             }
-            .tabItem {
-                Image("volume")
-            }
             .tag(Tab.home)
+            .background(TabBarReader{ tabBar in
```
Comment (Npikielny) [inline Volume/Views/MainView/MainView.swift:25]:
```suggestion
            .background(TabBarReader { tabBar in
```

File: Volume/Views/MainView/MainView.swift [hunk 13]
```diff
@@ -92,6 +131,34 @@ extension MainView {
...
+        var body: some View {
+            VStack(alignment: .center, spacing: 0) {
+                Spacer()
+                
+                icon
+                    .resizable()
+                    .aspectRatio(contentMode: .fit)
+                    .frame(width: size, height: size)
+                
+                Spacer(minLength: 4)
```
Comment (Npikielny) [inline Volume/Views/MainView/MainView.swift:148]: Did not know you could do that, very cool.

File: Volume/View Models/BigReadArticleRow.swift [hunk 14]
```diff
@@ -14,26 +14,24 @@ struct BigReadArticleRow: View {
...
+                    .resizable()
+                    .grayBackground()
+                    .frame(width: 120, height: 120)
+                    .padding(30)
             }
-            .frame(width: 180, height: 300)
+            ArticleInfo(article: article, showsPublicationName: true, largeFont: false)
         }
+        .frame(width: 180, height: 300)
         .accentColor(Color.black)
```
Comment (Npikielny) [inline Volume/View Models/BigReadArticleRow.swift:35]: Might be able to `.black` here

File: Volume/Views/BrowserView.swift [hunk 15]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+    }
+    
+    private var isArticleLoaded: Bool {
+        switch state {
+        case .loading:
+            return true
+        case .results(_):
+            return false
+        }
+    }
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:53]:
```suggestion
    private var isArticleLoaded: Bool { state == .loading }
```

File: Volume/Views/BrowserView.swift [hunk 16]
```diff
@@ -32,8 +34,82 @@ struct BrowserView: View {
...
+    
     // MARK: UI Components
```
Comment (Npikielny) [inline Volume/Views/BrowserView.swift:56]:
```suggestion
```

File: Volume/Views/MainView/MainView.swift (hunk 3 above)
Comment (Npikielny) [reply Volume/Views/MainView/MainView.swift]: Or debug print 👀

**Review by Npikielny — APPROVED** [review]
Looks great! Got a few nits here and there, but nice PR 😋.

File: Volume/Supporting Views/TabBarReader.swift [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+extension TabBarReader {
+    private class ProxyController: UIViewController {
+        var completion: (UITabBar) -> Void = { _ in }
+
+        override func viewWillAppear(_ animated: Bool) {
+            super.viewWillAppear(animated)
+            if let tabBarController = self.tabBarController {
+                self.completion(tabBarController.tabBar)
```
[the remark this answers is not in the document]
Comment (Npikielny) [reply Volume/Supporting Views/TabBarReader.swift:30]: selves (plural)

File: Volume/Supporting/Main.swift [hunk 18]
```diff
@@ -19,6 +19,7 @@ struct Main: App {
         configureFirebase()
         configureNotifications()
         UINavigationBar.appearance().backgroundColor = .white
+        UINavigationBar.appearance().tintColor = UIColor(Color.volume.orange)
```
[the remark this answers is not in the document]
Comment (dishanest) [reply Volume/Supporting/Main.swift:22]: This is the only place in the code where `UIColor(Color...)` is used, so probably not justified

File: Volume/Views/BrowserView.swift (hunk 4 above)
Comment (dishanest) [reply Volume/Views/BrowserView.swift:38]: It's not very declarative to manage state w/ optionals :/

File: Volume/View Models/BigReadArticleRow.swift (hunk 14 above)
Comment (dishanest) [reply Volume/View Models/BigReadArticleRow.swift:35]: went ahead and removed `Color.` everywhere else in the code where it isn't needed.

File: Volume/Views/BrowserView.swift (hunk 15 above)
Comment (dishanest) [reply Volume/Views/BrowserView.swift:53]: Tried this, it don't work :( `BrowserViewState<Results>` does not conform to `Equatable`, and I don't want to write code to try to make `Results` (which is a generic parameter) conform to `Equatable` too.

Comment (Npikielny) [reply Volume/Views/BrowserView.swift:53]: ohhhh that makes sense

**Review by NgaiJustin — APPROVED** [review]
looks good!

File: Volume/View Models/ArticleRow.swift [hunk 19]
```diff
@@ -21,21 +21,19 @@ struct ArticleRow: View {
     }
 
     var body: some View {
-        NavigationLink(destination: BrowserView(initType: .readyForDisplay(article), navigationSource: navigationSource)) {
-            HStack(spacing: 20) {
-                ArticleInfo(article: article, showsPublicationName: showsPublicationName, largeFont: false)
+        HStack(spacing: 20) {
+            ArticleInfo(article: article, showsPublicationName: showsPublicationName, largeFont: false)
```
[the remark this answers is not in the document]
Comment (dishanest) [reply Volume/View Models/ArticleRow.swift:25]: we don't show the publication name when the magazine cell appears on the publication detail page. that would be redundant information.

**Pull request merged**